### ISSUE TITLE
Analyzer: Support generalized void

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#30177]
This is the analyzer specific issue for #30176, which has the details.

**Comment by eernstg** [comment]
[This CL](https://codereview.chromium.org/2990703002/) introduces support for the syntactic extensions in the analyzer, but does not yet check for usages of values of static type `void`. So, apart from the usual subtype checks, they are still allowed everywhere (but it should be a compile-time error to use such a value in most situations, e.g., `print(v)` and `var y = v;` should be compile-time errors when `v` is declared as `void v = null;` ).

**Comment by eernstg** [comment]
The above-mentioned change was landed on July 27th as 9336e199fd7360ee229de5b4ebef5940aab266ec. This issue will be closed when the static checks have been implemented as well.

**Comment by bwilkerson** [comment]
@eernstg Does the informal spec (https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md) include descriptions of all of the places where there ought to be an error?

**Comment by eernstg** [comment]
(I'm pretty sure I answered this question several days ago, but I can't find the answer anywhere, so here we go again ;-).

Yes, the informal spec of generalized void is intended to contain a specification of all contexts where it is an error to evaluate an expression of type void. However, it's specified in a negated manner, as a list of cases where it's OK, and all other contexts are then the ones where an error should occur (because those other situations are the vast majority). 

The whitelist can be found in the [informal spec](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md) by searching for `situations:`.

The only other error is concerned with overriding; search for `overrides` to find that.

Note that there will be more errors in the future, when we introduce the notion of 'voidness preservation'. E.g., with voidness preservation we will almost certainly (it is not yet fully specified) have errors in the following situations:
```
List<Object> xs = <void>[]; // Voidness preservation violation.

void foo() {}
Object Function() f = foo; // Voidness preservation violation.
```
The general idea is that we will prevent some _indirect_ patterns where voidness is "forgotten".

But this is in the future, so it's currently only required to raise errors in the cases which are listed in the document.

**Comment by MichaelRFairhurst** [comment]
Given that this was reopened with the intention of supporting dart 2, that seems to indicate to me based on the informal spec that what we actually want to implement now is "voidness preservation."

Some other changes:

> the type void is considered to be equivalent to the built-in class Object in Dart 1.x, except when used as a return type, in which case it is effectively considered to be a proper supertype of Object. In Dart 2 subtyping, the type void is consistently considered to be equivalent to the built-in class Object.

> Also note that we may not allow return statements returning an expression of type void in Dart 2, but it is allowed here for backward compatibility.

> In Dart 2, it is a compile-time error when a method declaration D2 with return type void overrides a method declaration D1 whose return type is not void.

Also this:

> direct usage of the value of an expression of type void is a static warning (in Dart 2: an error)

Crucially, it looks like we're missing this:

> this document does not specify voidness preservation

and an anwer to whether return statements will still be allowed

**Comment by MichaelRFairhurst** [comment]
Ah, I see that I misunderstood. This ticket was never closed and reopened. I misread this:

@bwilkerson bwilkerson referenced this issue 22 days ago
 Closed
Support generalized void in the analyzer #31636 

I suppose that means, though, that we have two tasks here...to support generalized void checks at all, and then to support dart-2.0 generalized void checks.

**Comment by leafpetersen** [comment]
The voidness preservation is not part of this.  In order of priorities:

1) Allow `void` to be used syntactically as a type argument
2) Make sure subtyping/least upper bound work with types containing void
3) Make sure inference does the right thing
4) Add the limited syntactic checks specified in the informal proposal (*not* generalized void)

The first two I think are required to unblock https://github.com/dart-lang/sdk/issues/31869

**Comment by eernstg** [comment]
Right, I added support to the legacy front end of the analyzer for the generalized void syntax (cf. 1) in July, and `generalized_void_syntax_test.dart` succeeds now, so that as apparently working.

For subtyping (cf. 2), `void` is a top type, and that's clearly not the way it is treated. E.g., this causes a compile-time error in `main` of `void_type_usage_test.dart`, in the subtest `none` (that is, code which is always present), that shouldn't be there: "The argument type 'int' can't be assigned to the parameter type 'void'" --- yes it can! So that needs to be fixed.

(cf. 4) `void_type_usage_test.dart` serves to test that a lot of usages of expressions of type `void` are rejected as compile-time errors, or that they are accepted. However, that test fails in confusing ways. In particular, a number of test cases incur a compile-time error as expected, but because of the compile-time error in the code which is always present the intended compile-time error may or may not occur (and in some cases it certainly doesn't). Similarly, the always-present compile-time error makes all cases fail that have an `ok` expected outcome. In summary, the status file looks like this test is running almost successfully, but it's actually failing all over the place.

**Comment by MichaelRFairhurst** [comment]
Thanks for the info everyone. I think I have everything I need!

That said, I'll be taking a quick break on this today to get ready for dartconf. Hopefully its not an all-day break, but, so far the stuff that was supposed to be quick and easy there hasn't been!

**Comment by eernstg** [comment]
Please note the issues described in #31883: It explains why `void_type_usage_test.dart` may not easily be improved such that it avoids reporting a successful outcome even though we "got the wrong error". There are lots of other subtests in `void_type_usage_test.dart`, so I've marked it as P3, but you may wish to check manually that you actually get a "can't evaluate an expression of type void here" error rather than some other error, when implementing this feature.

**Comment by MichaelRFairhurst** [comment]
I also ran the quick test of supporting _isTop to include void. I'm seeing 6 failures in the dart analyzer test cases, which is really not many, so it should be easy to tackle this overall unless, say, the language tests have more failures or something.

**Comment by MichaelRFairhurst** [comment]
I also noticed some discrepancies between the internal spec and the language test:

```
void testVoidParam(void x) {      
  ...                
  return x;   //# param_return: compile-time error 
```

and the spec:

```
It is a static warning for an expression to have type void (in Dart 2: a compile-time error), except for the following situations:
...
In a return statement return e;, when the return type of the innermost enclosing function is the type void, e may have type void.
```

I'm assuming we should go with err on returning void values for now, that the language test here is what we want to go with rather than the informal spec.

**Comment by eernstg** [comment]
@MichaelRFairhurst Good catch! [This CL](https://dart-review.googlesource.com/c/sdk/+/34582) makes corrections such that all the existing return statements are compile-time errors (because the return types of `test..` functions are now `dynamic` rather than `void`), and a new test function has been added to test that return-void-in-void-function is ok.

**Comment by eernstg** [comment]
[Said CL](https://dart-review.googlesource.com/c/sdk/+/34582) has been [landed](https://dart.googlesource.com/sdk/+/bf5b9087d24157ebadd00177a028a3be24367a6a).

Note that the resulting `void_type_usage_test.dart` looks a lot like it has dead code (e.g., the new function `testReturnToVoid` has many consecutive return statements), but since this is a multi-test we won't actually have any dead code in each test. So there is no need to "fix that".

**Comment by MichaelRFairhurst** [comment]
Thanks Erik for the quick clarification!

**Comment by MichaelRFairhurst** [comment]
Also, is it really desirable that `as` on `void` is ok, and `is` isn't?

People will have to write

```
  void f() {
    final v = voidMethod();
    var resultAfterBranch;
    try {
      final secretlyReturnedVal = v as ReturnedVal;
      ...
    } on ..... {
      ...
   }
```

instead of

```
 void f() {
   final v = voidMethod();
   if (v is ReturnedVal) {
     ...
   } else {
     ...
   }
 }
```

My initial impression is that this is boilerplate, and perf hit, for no value. So, is this a slip-up, or is there value in making people go through these hoops?

**Comment by MichaelRFairhurst** [comment]
I guess I just realized; the examples have errors, right? The assignment doesn't fall into the case of where void may be used.

So I guess the reason we only allow `as` is that the `as` version can be rewritten to not use a temporary ie

```
void f() {
  try {
    final secretlyReturnedVal = voidMethod() as ReturnedVal;
    ...
  } on ..... {
    ...
  }
```

whereas the equivalent is not possible with `is` unless we allow void values to be assigned to void variables, and/or the user relies on the method being idempotent

```
  void f() {
    if(voidMethod() is ReturnedVal) {
      v = voidMethod as ReturnedVal; // better be idempotent!
    ...
```

Hmm. Well, nobody said designing a language was easy. Though this escape hatch definitely feels much more like an escape hatch than a feature, in this case.

**Comment by leafpetersen** [comment]
> Also, is it really desirable that as on void is ok, and is isn't?

Erik can comment definitively, but I believe the answer is that you really shouldn't be looking at `void` values to begin with - you're in some sense violating the implicit contract ("don't use the return value, it's arbitrary, and can change").  We allow it, but I don't think we need to make it easy.  If you really want to test the result, you can do `(voidMethod() as dynamic) is ReturnedVal` with presumably no perf hit.

**Comment by MichaelRFairhurst** [comment]
`voidMethod() as dynamic` ah yes, use a different top type to get out of voidness semantics. Nice.

**Comment by MichaelRFairhurst** [comment]
This is just about ready for review, but I ran into issues within ddc that I'm not sure are very "safe."

```
void myFun() {}

myOtherFun() => myFun();
```

results in 'the return type void isn't a dynamic, as defined by myFun'.

This begs the question; does this part of the spec have to be done now?

```
In Dart 2, it is a compile-time error when a method declaration D2 with return type void overrides a method declaration D1 whose return type is not void.
```

Though, the problem may be in the inference rather than in this rule.

**Comment by leafpetersen** [comment]
Can you elaborate the example a bit?  Bleeding edge DDC compiles and runs this program just fine:

```dart
void myFun() { print("hello");}

myOtherFun() => myFun();


void main() {
  myOtherFun();
}
```

I'm not sure I understand the reference to the override error.

**Comment by MichaelRFairhurst** [comment]
for full info, this is the current state of the CL: https://dart-review.googlesource.com/c/sdk/+/35002

**Comment by MichaelRFairhurst** [comment]
After the void changes, that is no longer valid and it requires an explicit "void" return type annotation on "myOtherFun".

I don't think I touched inference in any ways that affect that, though I will double check.

The only other failing code that I saw was cases where it indeed was using a void value, which previously was only a hint, in package:expect. In that case I used the `as dynamic` escape hatch. If I had changed the type of the corresponding typedef to `() -> dynamic` from `() -> void`, then I was worried that occurrences of specifying that function would have problems overriding the dynamic return type to be void, as in the DDC case at inference time.

**Comment by leafpetersen** [comment]
> but I ran into issues within ddc that I'm not sure are very "safe."

Maybe I'm misunderstanding. Are you saying that you ran into issues with the DDC source code, or with running DDC tests, or... ?

I made some comments on the CL.

**Comment by MichaelRFairhurst** [comment]
> Are you saying that you ran into issues with the DDC source code

I wasn't sure! I think now, the answer is no.

So not only does the example I gave (`myOtherFun() => myVoidFun()`) currently work on bleeding edge, but it *should* work after the void changes are landed, right?

Definitely seeing lots and lots of language tests that don't like this change, so I'm assuming its a bug in my CL.

**Comment by MichaelRFairhurst** [comment]
Ah, this is way simpler than I realized.

This rule is too strict, at least for now:

> In a return statement return e;, when the return type of the innermost enclosing function is the type void, e may have type void.

In the case provided, `myOtherFun` is typed as `() -> dynamic`. There is no type inference here. Correct?

Unfortunately, that means it can't `=>` return a `void` result, such as `myVoidFun()`.

So we can either special case arrows, or `dynamic`/`void`, or maybe there is some inference here that I don't know about (or should be).

**Comment by MichaelRFairhurst** [comment]
Or fill in the type ie `void myOtherFun() => myVoidFun()` in all cases, which appears to be a lot.

**Comment by leafpetersen** [comment]
So if I understand you correctly: the code that has these errors (that is, the code being analyzed) is in language tests?

I think @eernstg and @lrhn should comment on whether there is an intended exception for either `dynamic` contexts or for `=>` functions, but as I read the proposed generalized void rules, that code should be an error.

Is it possible to land this incrementally?  In the priority list that I gave above, I believe that these errors fall under 4.

**Comment by MichaelRFairhurst** [comment]
> Is it possible to land this incrementally? In the priority list that I gave above, I believe that these errors fall under 4.

Yes, incrementally seems like the right call here.

**Comment by MichaelRFairhurst** [comment]
The errors in the language tests are here: https://logs.chromium.org/v/?s=dart%2Fbuildbucket%2Fcr-buildbucket.appspot.com%2F8957199686594153152%2F%2B%2Fsteps%2Fanalyze_tests_checked%2F0%2Fstdout, many or most of them are this specific "returning null in a dynamic function" problem.

I will ask Keerti to check this against google3, too, see if its only the tests that are affected or if its a common pattern/mistak.

**Comment by eernstg** [comment]
I'll respond to a few things that I noted in the discussion above; I'm not 100% sure which things have been resolved already (or how ;-), so bear with me if there's a little bit of redundancy.

> Also, is it really desirable that as `on void` is ok, and `is` isn't?

Both `on void` and `is void` are errors (they are not in the list of exceptions).

> `final v = voidMethod();`

That would be an error (assuming that `voidMethod` has return type `void`) because it evaluates an expression of type `void` in a context that is not one of the six exceptions. But you could "step out of voidness" immediately with `final v = voidMethod() as ReturnedVal;`, or you could allow for a broader range of values with `final v = voidMethod() as Object;` and then later use `v is T` for various `T` as needed. Note that `as Object` may be a no-op at run time, so there is no performance hit; and `v is ReturnedVal` is still in your "instead of" example, so that cost would be paid anyway.

In any case, casting-out-of-void is intended to be a very, very special case. You can do it if you're in trouble (because of legacy code in a third party library or whatnot), but in general it shouldn't count because it doesn't happen.

The [next comment](https://github.com/dart-lang/sdk/issues/30177#issuecomment-358058419) has this:
```dart
void f() {
  if(voidMethod() is ReturnedVal) {
    v = voidMethod() as ReturnedVal; // better be idempotent!
    ...
```
In this case you can't use promotion (because promotion only applies to variables, not to non-trivial expressions like function calls), but you can cast out of void in one step, and then promote:
```dart
void f() {
  var v = voidMethod() as Object;
  if(v is ReturnedVal) {
    // `v` has type `ReturnedVal` here, and evaluation happened only once.
    ...
```
Again, it doesn't matter so much if this is slightly verbose and inconvenient, because it just shouldn't happen. It's an emergency-only feature to cast out of void.

```dart
void myFun() {}
myOtherFun() => myFun(); // (*)
```
The line marked (*) is a compile-time error because the expression `myFun()` is evaluated and has type `void`, and the situation is not one of the six exceptions where that is allowed. I can see that there was some doubt on this:

> it requires an explicit "void" return type annotation on "myOtherFun".
> ... [but it] should work after the void changes are landed, right?

No, `void myOtherFun() => myFun();` would be OK because it's exception 6, but `myOtherFun() => myFun();` is an error.

**Comment by lrhn** [comment]
> but `myOtherFun() => myFun();` is an error.

It is only an error if we don't infer `void` as the return type, which we should since the body expression has type `void`. We cannot talk about what the return type of a function is until its return type has been inferred, which is why type inference must be one of the first things to happen.

I have one rule that I insist on: It shouldn't make any difference whether a type is inferred or written explicitly. (Or, more precisely, if you take a program, infer types, and then insert all the inferred types explicitly, then the program should do exactly the same). So, I don't want a rule that differs whether `void` is written or inferred as return type.

**Comment by eernstg** [comment]
Right, I was assuming that `myOtherFun` and `myFun` are library functions (top level), in which case I do not think any inference would be applicable to fill in missing type annotations in the signature. So `myOtherFun` would have return type `dynamic`.

For a function literal we could infer the return type using the type-from-context, and for an instance method we could obtain the return type from a declaration in a superinterface of the enclosing class. Granted, for a situation where a function literal is used in a context that provides no input on the desired return type, we would use the return expression, and that would make `() => myFun()` an expression of type `void Function()`.

**Comment by MichaelRFairhurst** [comment]
It seems to me we may want to allow

x ? voidValue : voidValue;

I only saw this once, but it seems to me that the standard rules would then apply -- the ternary would have a void type due to LUB even if only one of the cases results in a void type. It would then be an error to pass that into functions/assign its result, etc etc.

Makes it legal to do like

```
void increment() ...
void throwException() ...
void foo() => isLocked() ? throwException() : increment();
```

**Comment by eernstg** [comment]
The language team discussed allowing `boolValue ? e1 : e2` where `e1` or `e2` (or both) have static type `void` (and the conditional expression as a whole would then also have static type `void`, which would prevent it from occurring in all sorts of "normal" expressions). But after a (loud ;-) discussion the majority decided against it. The main point was that we wanted to keep the set of locations where a void expression can occur as minimal as possible while still allowing it in the locations where the value is directly discarded, plus the cast for emergencies. With `?:` we would go beyond this rationale because there is an implied computational step whereby "the void-typed value is passed through the branching point of the conditional operator". So it lets "the void value" take one step into compute-land, and we only wanted to allow it to live in discard-land. ;)

**Comment by MichaelRFairhurst** [comment]
Should these no longer be compile time errors?

```
void badReturnTypeAsync() async {} // //# 01: compile-time error                 
void badReturnTypeAsyncStar() async* {} // //# 02: compile-time error            
void badReturnTypeSyncStar() sync* {} // //# 03: compile-time error 
```

I can see both ways. Its sound, after all. And on the one hand, it essentially lets `void` do this:

```
  void myVoidAsyncFun() => myAsyncFun();
  Future<void> myAsyncFun() async { ... }
```

in a single keyword. On the *other* hand, the above code is a very intentional way of disallowing api consumers of doing a very normal thing. Maybe its *good* that that's not easy to do on accident.

This should probably at the very least get paired with a lint "async_functions_prefer_future_void_to_void" which suggests "did you mean to return Future<void> instead of void from your async function so that its awaitable?"

**Comment by MichaelRFairhurst** [comment]
Only just noticed that Erik had responded to assert that `x ? voidVal : voidVal` was *explicitly* decided against.

I guess in this case, this code in dart2js *should* be flagged. There's nothing in my CL that suddenly causes it to infer `void`. Its just that before it was not flagged for being an operand of the ternary.

A lot more breaking changes in here than I expected; coordinating this will be tricky. I will likely have to start with Hints to ease the transition.

I suppose I could try writing a dartwing migration! @mk13 a lot in here, but here are the major issues I'm seeing for migration purposes:

* change `cond ? voidValue : voidValue` to `if (cond) { voidValue; } else { voidValue; }`
* change `foo() => voidValue;` to `void foo() => voidValue;`
* (this one, we may specifically *not* want to do, but just for completeness' sake) change things like `func(voidValue)` to `func(voidValue as Object)`

Dartwing should be able to do all three of these things pretty straightforwardly, right?

**Comment by MichaelRFairhurst** [comment]
@leafpetersen @eernstg is it a good idea to land this change as simply exposing `void` as `Top`, maybe with hints for where its used?

Upsides:
* won't break the roll
* unblocks #31869 (right?)
* entirely backwards compatible (in theory, since it makes the type system *more* accepting...right?)
* (if we land w/ hints) teams can see & fix issues, which really are issues, before its a breaking change

Downsides:
* Using void types incorrectly will no longer be a sidecast, so it can be used anywhere (BAD)
* Not all teams pay attention to hints. Some have too many hints already to pay attention.

The two downsides *do* mean that while some teams will be able to slowly clean up their code, others will be vulnerable to a slip in code quality, soundness, and/or productivity.

Perhaps if the migration on the whole is quick, unblocking others makes this a net win. But the sidecast issue is a huge one.

**Comment by leafpetersen** [comment]
I'm concerned about biting off another migration here, and I'm also concerned that if we're seeing a lot of uses of a particular code pattern, that might be a sign that we're not on the right track by disallowing it.  For example, perhaps we should extend the exception on returning `void` things to `dynamic` as well?  And if a *lot* of users are using conditional expressions returning `void`, then maybe that's a sign that it's a useful pattern?

I think the highest priority is making `void` properly useful as a top type.  I would suggest landing that as soon as possible.  If you also prepare a CL that issues the specified warnings, then I (or you) can test it on internal code to better understand the magnitude of the migration, and what kind of code patterns we would be outlawing.

**Comment by lrhn** [comment]
I'm sure that there are uses of the conditional operator as a short-hand for an `if` statement. I just didn't want to encourage that. It's hard to read - using an expression means that the reader will expect it to return a value, and they have to study it closely to realize that it's just "returning void".
(That's also my opposition to allowing void-arrow functions).

If `if` statements are too cumbersome to write, or if braces are too cumbersome to write, we can improve the syntax. We shouldn't encourage using an expression where a statement is intended, that moves the burden from the person writing (not so many) characters to the person who has to read the underlying concepts from the syntax.

That is, I'm very, very happy making `=> condition ? voidExpression() : null` invalid. It's much more readable as:
```dart
void assertCondition(bool condition, Object arg) { 
  if (_isConsoleDefined) JS('void', 'console.assertCondition(#, #)', condition, arg);
}
```

As for allowing `void` as return type on `async`, `async*` and `sync*` functions: I *definitely* want it for `async`, but it makes less sense for `async*` and `sync*`  because the function does *nothing* if you don't use the returned value, so it's most likely an error.

(I'm all for a gentle migration experience, as long as we end up being fully migrated.)

**Comment by eernstg** [comment]
> Should these no longer be compile-time errors? 
```dart
void badReturnTypeAsync() async {} // //# 01: compile-time error
void badReturnTypeAsyncStar() async* {} // //# 02: compile-time error
void badReturnTypeSyncStar() sync* {} // //# 03: compile-time error
```
The [feature spec of generalized void](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md) does not address this case.

The language specification (`dartLangSpec.tex`) has not been updated with respect to the generalization of void, which means that some sections will need to be changed a lot (e.g, `\ref{typeVoid}`).

Still, the following verbiage in the language specification could be relevant, and might survive (with warnings changed to errors):
```
It is a static warning if the declared return type of a function marked \ASYNC{}
  is not a supertype of \code{Future<\mbox{$T$}>} for some type $T$.
It is a static warning if the declared return type of a function marked \SYNC*
  is not a supertype of \code{Iterable<\mbox{$T$}>} for some type $T$.
It is a static warning if the declared return type of a function marked \ASYNC*
  is not a supertype of \code{Stream<\mbox{$T$}>} for some type $T$.
```
But `void` is now a top type, so the given declarations are no longer compile-time errors (`T` can be an arbitrary type and it will always will work).

Unless we specifically decide that they should be errors, and change the spec accordingly.

If the bodies are not empty, the following may also be relevant:
```
It is a static type warning if the body of $f$ is marked \ASYNC{} and the
type \code{Future<$flatten(T)$>} (\ref{functionExpressions}) may not be
assigned to the declared return type of $f$.
Otherwise, it is a static type warning if $T$ may not be assigned to the declared return type of $f$.
...
It is a static type warning if either:
\begin{itemize}
\item the body of $f$ is marked \ASYNC* and the type \code{Stream<T>}
  may not be assigned to the declared return type of $f$.
\item the body of $f$ is marked \SYNC* and the type \code{Iterable<T>}
  may not be assigned to the declared return type of $f$.
\end{itemize}
```
Here, the type `T` is the static type of the expression `e` in a return statement `return e;` which is under scrutiny. Given that we have no special rule against assignments to a `void` typed variable (we can do `void x = 42` now), and given that `void` is a top type, this would be interpreted to mean that those declarations are no longer compile-time errors.

It is again a consistent choice to allow these kinds of functions.

However, there is a "moral" equivalent here which says the opposite. I included the sentence `Otherwise, ..` above, because it's _the_ place where it is specified that it is an error to do `return e;` in a "normal" function (not `async` etc.), unless the type of `e` is assignable to the return type. So we might think that these declarations would be compile-time errors if something like the following declarations are compile-time errors:
```dart
void badReturnTypeAsync() async {
  return new Future.value(null);
}

void badReturnTypeAsyncStar() async* {
  return new Stream.empty();
}

void badReturnTypeSyncStar() sync* {
  return new [];
}
```

These declarations have been flagged with a static type warning in Dart 1, based on the criterion that the returned expression should have a type which is assignable to the return type, and in Dart 1 nothing other than `dynamic` is assignable to `void`.

In Dart 2, everything is assignable to `void`, but in the language team we have had many strong expressions of a desire to ensure that `return e;` will remain an error in a function with return type `void` when `e` has a static type different from `void`.

The argument has always been that this is error prone: The developer wishes to return something (presumably something meaningful), but far away there is a declaration of a `void` return type which declares to the world that the returned value should be discarded.

There is no support in the [feature spec](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md) for these errors, and we've seen that the current wording in the language specification will make it a non-error, but I suspect that the language team will be quite happy about deciding this.

So here's a [CL on the feature spec](https://dart-review.googlesource.com/#/c/sdk/+/35680) to consider, Lasse and Leaf.

**Comment by eernstg** [comment]
> As for allowing void as return type on async, async* and sync* functions: I definitely want it for async, ..

@lrhn, why was that? The [CL](https://dart-review.googlesource.com/#/c/sdk/+/35680) mentioned above makes it an error.

**Comment by eernstg** [comment]
If we end up adding exceptions such that `b ? e1 : e2` are allowed even when `e1` or `e2` has type `void` (and then giving the whole expression the type `void`, such that we can't, say, put it into bigger expressions like `foo(true ? print('Hello, voidness!') : null)`), here's a [CL](https://dart-review.googlesource.com/c/sdk/+/35740) which adds support for that.

**Comment by MichaelRFairhurst** [comment]
I've asked Keerti to help me find the scope of these breakages. 

> I think the highest priority is making void properly useful as a top type. I would suggest landing that as soon as possible.

@devoncarew @leafpetersen [This](https://dart-review.googlesource.com/c/sdk/+/35520) does nothing but support void as Top. It has [248](https://logs.chromium.org/v/?s=dart%2Fbuildbucket%2Fcr-buildbucket.appspot.com%2F8957109130880293072%2F%2B%2Fsteps%2Fanalyze_tests%2F0%2Fstdout) failing language tests due to being far too permissive. Maybe that's not a problem, but, that makes me think that the easiest rollout strategy is:

* undo the inference change so that void not being top isn't a huge issue
* roll out hints for everywhere that void will not be able to be used
* let users fix those hints themselves as much as possible (this should improve their code, too)
* upgrade the void usages to a static warning
* ^ this stage can perhaps skip ternaries, returning void in dynamic, if those migrations are large/ongoing
* flip void to Top and reroll the inference changes

**Comment by MichaelRFairhurst** [comment]
> I think we're at best going to have explicitly suppress some error messages (e.g when returning void in a dynamic context)

Agreed. The suppressable parts are:
* returning void in a dynamic context (current approach)
* using warnings (current approach -- so StaticWarningCodes don't affect internal users?? That'd be great!)
* using hints: too loose for the language tests to be happy where it expected void couldn't be used since it used to be a side-cast and isn't any longer
* ternaries -- I think this is a good idea to suppress for migration purposes

Currently, most disallowed usages of void are related to it simply not having the getters/methods/operators. These therefore just don't exist ie `voidFn() + 2` and the language tests are happy with the new errors in these cases.

The next most common currently rejected usage of void is side-casts, ie, `takeInt(voidFn())`, which no longer fail once `void` becomes a form of Top, since it becomes a simple upcast. The language tests don't like the lack of an error here.

We can then disallow void usage in this case by saying that void cannot be used as an argument with the new warning code USE_OF_VOID_RESULT.

However, that's not fully backwards compatible either, because not all usages of void involved side-casts -- some were upcasts from void to Top. This creates problems where people did things like `print(voidFn())`, which you would think are rare, but definitely occurred in the language tests. The result of trying the CL will tell us how rare this really is.

It does go beyond arguments & things that were rejected due to missing methods, too. That's where we get `x ? voidFn() : voidFn()` as the prime example, but there are others such as `x ?? voidFn()` and 'void x = voidFn()`. These don't result in any erroneous usages of void, by themselves.

Maybe we can attempt to catch and alert all downcasts `x = voidFn()` or `useInt(voidFn())` for all cases where the cast is not to Top, which I think would be the most backwards compatible option. However, this is essentially not making `void` Top, is it not? We'd have to be careful to do it in such a way that lets `Future<void>` be a supertype of `Future<int>` without letting `int x = voidFn()` work via implicit downcast. We would have to make sure that those checks go through different paths that are reliably used since the idea of having different subtyping rules in these cases would be new.

**Comment by MichaelRFairhurst** [comment]
@eernstg what exactly makes `for (x in <void>[]) {}` an error?

Using `void` as a type argument in a list literal?

Iterating over an `Iterable<void>`?

Both of these have different correlated edge cases (ie, `<void, void>{}` vs `await for (var x in new Stream<void>())` or however that stream/for stuff works again).

**Comment by lrhn** [comment]
I'm not sure we have any current rule that prohibits `for (var x in <void>[]){}`. It would make any *use* of `x` invalid, so it's unlikely to be useful.

We could argue that the code is equivalent to `var iterator = <void>[].iterator;while(iterator.moveNext()) { void x = iterator.current; .... }`, and that use of `iterator.current` is invalid, but I'm not sure it's worth the exception. 
On the other hand, I do want to prohibit `for (Object x in <void>[]) {}`, so we probably need a rule against an `Iterable<void>` in `for-in` and a `Stream<void>` in `await-for-in`.

**Comment by eernstg** [comment]
It's not completely obvious whether there are any rules against `for (var x in <void>[]) {}`. Definitely there is no error in having a `<void>[]` (but it's not going to be very useful).

Iterating over an `Iterable<void>` is more questionable. As Lasse mentioned, the given for-in statement 'is equivalent' to a specific while statement (language specification section 'For-in'), and that while statement does contain an error when `iterator.current` has static type `void`.

I'd say that this is already enough to say that it is an error (otherwise it is not equivalent), and I wouldn't really think of it as an exception.

But it would presumably be useful to make it explicit in that section For-in, as a `\commentary` sentence.

The same reasoning would apply to await-for-in, but in that case there is no `equivalent` based definition in the language specification, so we would need to mention specifically that a stream whose type at `Stream` is `Stream<void>` (that is, some subtype of `Stream` such that `Stream<void>` is a superinterface) is a compile-time error.

I'll create a CL to adjust generalized-void.md to say that, and then we can use that to get a decision.

**Comment by eernstg** [comment]
Here's [the CL](https://dart-review.googlesource.com/c/sdk/+/35920) that makes `for (var x in <void>[]) {}` and its ilk errors. @lrhn and @leafpetersen: WDYT?

**Comment by MichaelRFairhurst** [comment]
Another quirk I'd like to point out is that this is technically valid:

```
new Future<void>.of(null).then((x) => x);
```

This isn't quite the semantics I'd intuitively expect; and I think I have a contrived way to force an unintuitively uncaught bug

(note: I'm assuming voidness preservation is implimented here. Obviously, lack of voidness preservation can create bugs)

```
T foo<T>(T x) => ...;
new Future<void>.of(null).then(foo);
```

Add in a standard upcast...

```
// a method that looks like a getter, but performs a side effect and returns void
int theInts = null;
Future<void> getInts() async {
  theInts = await ...;
}

// a generic function
List<T> toList<T>(T t) => new List<T>()..add(t);

// lift Future<void> to Future<List<void>>
getInts().then(toList)
  // and pass an int into (void) -> bool, which is valid
  .then((r) => r.contains(4)));
```

Granted, this now has very little if anything to do with my original "quirk I'd like to point out."

I'll write here what the code *should* have looked like in the final example:

```
// load the global
getInts().then((_) =>
  // use the global
  theInts.contains(4)
);
```

Yeah, so looks like the nugget of the problem here would be the lifting of `Future<void>` to `Future<List<void>>`. I'm not sure the way around this one.

But then again, maybe that lift isn't the problem, maybe the problem is a passing non-void values into void parameters? That could be a hint or a lint. (it would have to also handle, likely, passing a List<int> into a List<void>, too...)

Winding back to the original "quirk", its specifically leveraging the fact that void can be returned in a void function. I guess it could sort of, combine with this, if you made an identity function that had side-effects:

```
getInts().then((x) {
    print("I got the ints!"); // side-effect
    return x; // return identity...indicative that users thought the value was useful?
  })
  .then(toList) // ok now we lift
  // and get the same problems as before
  .then((r) => r.contains(4)));
```

Seems like, unless I'm just missing a clever enough example, the ".then((voidValue) => voidValue)` identity problem isn't really a problem (its not the source of the bug here), but still not necessarily ideal.

Nobody said designing a type system was easy; there will be some corner cases to any solution. Still, I'm trying to see ahead of the curve. This *may be* getting kinda beyond any practical waters, and still figured I'd share.

**Comment by lrhn** [comment]
> Another quirk I'd like to point out is that this is technically valid:
> ```dart
> new Future<void>.value(null).then((x) => x);
> ```

It is, because having a void expression is valid at the end of a `=>` arrow if the function returns `void`, and here the inference will infer `void` for both `x` and the return type of the `(x)=>x` function,. The resulting future's type is `Future<void>`.

...
>Yeah, so looks like the nugget of the problem here would be the lifting of `Future<void>` to `Future<List<void>>`. I'm not sure the way around this one.

Me neither. The moment `void` is represented by a type variable, it really means `Object`, so there will be no runtime checks to catch usages. You can skirt around the `void` type and never have a static `void` variable or expression, and that's a consequence of allowing `void` as a type parameter at all.
We could say that a type variable that can be void (any type variable with no bound) must be treated as if it is potentially `void`, but that would be horrible to work with in practice.

What we really have is "void erasure" - it's checked statically *where possible* , but at runtime it's just `Object`. Your code is clever - it uses a generic function to go from `void` to `List<void>`, which is statically type-safe and dynamically unchecked. 

We could also just remove `void` from the language and use `Null` instead everywhere. it would be reasonable in a functional language (`Null` is the unit type after all, and, e.g., SML uses `unit` as the return type when it has no value to return). We didn't, mostly because it would be breaking and unexpected.

So, instead we went with static void + erasure, and we only promise a *best effort* at detecting void violatons. It's not a security feature, just a way to detect something that is probably an error.

**Comment by MichaelRFairhurst** [comment]
But yeah, I'm likely shaking up the nest for no reason. The implementation of void here is great and I love the simplicity. Little quirks are expected in any complex system.

For solutions not related to the void typesystem.......I suppose that if we knew the variance of `X` in `T<X>`, we could say that only contravariant (not covariant or invariant) generics should be able to be applied by `void`. This makes `Future<void>` a valid type, but not `Sink<void>` (because why?) nor `List<void>` (and therefore not `List<Future<void>>`).

**Comment by lrhn** [comment]
Seems reasonable ... except that `Future<void>` is covariant and a primary reason for allowing `void` as a type argument in the first place. Another example was `Listener<void>`, where the type variable is used as return type for a generic listener pattern implementation.
It really is cases where we want to generalize over return types that made us allow `void` as a type parameter, and all the other uses of it are just unavoidable complications on top of that ... and return values are covariant.

We *want* `Future<void>` because it really is the correct type for an asynchronous void function. If we have that, we *need* to allow `(void x) { ... }` as argument to `Future.then`. Then we need to be able to handle `void` typed variables and expressions everywhere. And with inference, the moment we have void somewhere, it can end up anywhere. 

We don't particularaly *want* a `void` collection like `List<void>`. It's just that we can't have the things we want without also allowing that.

**Comment by eernstg** [comment]
In [this comment](https://github.com/dart-lang/sdk/issues/30177#issuecomment-358694328) I mentioned that I had created a CL that allows conditional expressions to contain branches of type void (so we can have `void foo() => b ? e1 : e2;` where `e1` and `e2` can have the type void).

Proposals for an extended version of the same idea have come up since then, such that we would generally allow data flow where the value of an expression of type void is allowed to flow to a sink of type void (that is, be assigned to a variable of type void, passed as an actual argument where the corresponding statically known formal parameter has type void, returned in a void function). Here's a CL for that: https://dart-review.googlesource.com/c/sdk/+/35740. As mentioned in that CL, it subsumes the [above-mentioned CL](https://dart-review.googlesource.com/c/sdk/+/35740) about conditionals and we would only land at most one of them.

**Comment by MichaelRFairhurst** [comment]
some remaining todos:

[ ] land the CL that makes usages of void types errors, for a starting point that works internally
[ ] triage FutureOr<void>
[ ] update analyzer to be fully compliant with the spec, confirm passes in g3
[ ] update internal/external code to no longer return void in dynamic functions
[ ] update analyzer/spec to disallow returning void in dynamic functions
[ ] experiment with making void the only top type

**Comment by MichaelRFairhurst** [comment]
@leafpetersen @eernstg @devoncarew I think this is done enough for dart2/the purpose of this ticket. WDYT?

Much of the rest is spec work, and the void return case which could probably be its own ticket, from what I can tell.

**Comment by leafpetersen** [comment]
The main issue will be flutter.  Anything remaining will have to be run through their breaking change process, which means breakage will have to be fairly limited and well-justified.

**Comment by eernstg** [comment]
> Much of the rest is spec work ..

That would belong to #30597 (which lists all the relevant feature spec CLs, including the ones mentioned here), and it should be possible to handle this issue independently. For the record, I've added a comment on #30597 noting that some discussions here are relevant to said CLs.

**Comment by MichaelRFairhurst** [comment]
Assuming `FutureOr` and Top changes are part of #30597, I made #32161 to address the work I'm still looking at as I have availability. This ticket can now be closed

**Comment by leafpetersen** [comment]
void_type_function_types_test/none does not parse.  Small repro which shows the incorrect parse error:

```dart
typedef G2 = void Function([void]);
```

**Comment by leafpetersen** [comment]
Also failing:
void_type_usage_test/param_boolean_and_right: MissingCompileTimeError
void_type_usage_test/param_boolean_negation: MissingCompileTimeError
void_type_usage_test/param_boolean_or_right: MissingCompileTimeError

**Comment by leafpetersen** [comment]
Also failing:
void_type_usage_test/param_use_in_string_interpolation: MissingCompileTimeError # https://github.com/dart-lang/sdk/issues/30177

**Comment by MichaelRFairhurst** [comment]
Closing. All tests are either passing except the issues filed in #35508, #34319, both of which are unrelated to generalized void.
